### ISSUE TITLE
HIP runtime error when targeting both CUDA and HIP

### ISSUE THREAD

**Issue by al42and** [AdaptiveCpp/AdaptiveCpp#830]
**Bug summary**

When the app is compiled for both CUDA and HIP (with explicit multipass), the embedded HIP objects cannot be found at run time:

```console
$ syclcc -O3 --hipsycl-explicit-multipass --hipsycl-targets='hip:gfx1032;cuda:sm_86' test.cpp -o test && ./test
clang: warning: CUDA version is newer than the latest supported version 11.5 [-Wunknown-cuda-version]
Trying to run on NVIDIA GeForce RTX 3060
   Done!
Trying to run on hipSYCL OpenMP host device
   Done!
Trying to run on    ## HIP reports empty name for AMD GPUs 
[hipSYCL Error] from /home/aland/hipSYCL/src/runtime/hip/hip_code_object.cpp:125 @ build(): hip_executable_object: could not create module (error code = HIP:218)
[hipSYCL Error] from /home/aland/hipSYCL/include/hipSYCL/runtime/kernel_cache.hpp:291 @ get_or_construct_code_object_impl(): kernel_cache: code object creation has failed
[hipSYCL Error] from /home/aland/hipSYCL/src/runtime/hip/hip_queue.cpp:522 @ submit_kernel_from_code_object(): hip_queue: Code object construction failed
   Fail!

```

The error is distinct from simply not having a target:

```console
$ syclcc -O3 --hipsycl-explicit-multipass --hipsycl-targets='cuda:sm_86' test.cpp -o test && ./test
clang: warning: CUDA version is newer than the latest supported version 11.5 [-Wunknown-cuda-version]
clang: warning: CUDA version is newer than the latest supported version 11.5 [-Wunknown-cuda-version]
Trying to run on NVIDIA GeForce RTX 3060
   Done!
Trying to run on hipSYCL OpenMP host device
   Done!
Trying to run on 
[hipSYCL Error] from /home/aland/hipSYCL/include/hipSYCL/runtime/kernel_launcher.hpp:86 @ find_launcher(): No kernel launcher is present for requested backend
[hipSYCL Error] from /home/aland/hipSYCL/src/runtime/hip/hip_queue.cpp:307 @ submit_kernel(): Could not obtain backend kernel launcher
   Fail!
```

When targeting only the AMD device, it works fine:

```console
$ syclcc -O3 --hipsycl-explicit-multipass --hipsycl-targets='hip:gfx1032' test.cpp -o test && ./test
Trying to run on NVIDIA GeForce RTX 3060
[hipSYCL Error] from /home/aland/hipSYCL/include/hipSYCL/runtime/kernel_launcher.hpp:86 @ find_launcher(): No kernel launcher is present for requested backend
[hipSYCL Error] from /home/aland/hipSYCL/src/runtime/cuda/cuda_queue.cpp:314 @ submit_kernel(): Could not obtain backend kernel launcher
   Fail!
Trying to run on hipSYCL OpenMP host device
   Done!
Trying to run on 
   Done!
```

**To Reproduce**

Commands are listed above. I think any code will suffice, but here is what I used.

```cpp
#include <SYCL/sycl.hpp>
#include <iostream>

void run_kernels(const sycl::device &syclDevice) {
  unsigned int numThreads = 512;
  sycl::queue queue = sycl::queue(syclDevice);
  int *ptr = sycl::malloc_device<int>(numThreads, queue);
  queue
      .submit([&](sycl::handler &cgh) {
        cgh.parallel_for<class KernelA>(sycl::range<1>{numThreads},
                                        [=](sycl::id<1> id) { ptr[id] = 0; });
      })
      .wait_and_throw();
  sycl::free(ptr, queue);
}

int main() {
  for (const auto &dev : sycl::device::get_devices()) {
    std::cout << "Trying to run on " << dev.get_info<sycl::info::device::name>() << std::endl;
    try {
      run_kernels(dev);
      std::cout << "   Done!" << std::endl;
    } catch (std::exception) {
      std::cout << "   Fail!" << std::endl;
    }
  }
}
```

**Expected behavior**

Code successfully runs on devices it was compiled for.

**Describe your setup**

* hipSYCL 989026577b62409139c856491e9d67707448780a. A quick test shows that the problem is also reproduced with 30674861128c93e0a14c8a2a35fb84c8f68e91f4.
* Describe the dependencies that hipSYCL sits on top of in your setup: 
   * clang++ 15 from apt.llvm.org (`15.0.0-++20220914063142+6fe69891d15c-1~exp1~20220914063154.53`)
   * CUDA 11.7 from https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/
   * ROCm 5.2.0 from https://repo.radeon.com/rocm/apt/debian/

**Optional additional diagnostic information**

Looking at the flags used for linking, nothing is clearly suspicious. Note: changing the order of the targets changes the flags a bit.

```console
$ HIPSYCL_DRYRUN=1 syclcc -O3 --hipsycl-explicit-multipass --hipsycl-targets='hip:gfx1032;cuda:sm_86' test.cpp -o test
/usr/lib/llvm-15/bin/clang++ -I/home/aland/local/bin/../include/ -D__HIPSYCL__ -std=c++17 -isystem /home/aland/local/bin/../include/hipSYCL/std/hiplike -isystem /usr/lib/llvm-15/lib/clang/15.0.0/include/.. -U__FLOAT128__ -U__SIZEOF_FLOAT128__ -I/opt/rocm//include -I/opt/rocm//include --rocm-device-lib-path=/opt/rocm//amdgcn/bitcode --rocm-path=/opt/rocm/ -fhip-new-launch-api -mllvm -amdgpu-early-inline-all=true -mllvm -amdgpu-function-calls=false -D__HIP_ROCclr__ -x hip -D__HIPSYCL_ENABLE_HIP_TARGET__ -D__HIPSYCL_CLANG__ -fplugin=/home/aland/local/bin/../lib/libhipSYCL_clang.so --cuda-device-only --cuda-gpu-arch=gfx1032 -c -o /tmp/hipSYCL-syclcc-9ypnu18w/hipfb-gfx1032.hipsycl_kernels -fpass-plugin=/home/aland/local/bin/../lib/libhipSYCL_clang.so -O3 test.cpp
/usr/lib/llvm-15/bin/clang++ -I/home/aland/local/bin/../include/ -D__HIPSYCL__ -std=c++17 -U__FLOAT128__ -U__SIZEOF_FLOAT128__ -isystem /home/aland/local/bin/../include/hipSYCL/std/hiplike -x cuda -D__HIPSYCL_ENABLE_CUDA_TARGET__ -D__HIPSYCL_CLANG__ -fplugin=/home/aland/local/bin/../lib/libhipSYCL_clang.so --cuda-path=/usr/local/cuda --cuda-device-only --cuda-gpu-arch=sm_86 -S -o /tmp/hipSYCL-syclcc-9ypnu18w/nvptx-sm_86.hipsycl_kernels -fpass-plugin=/home/aland/local/bin/../lib/libhipSYCL_clang.so -O3 test.cpp
/usr/lib/llvm-15/bin/clang++ -I/home/aland/local/bin/../include/ -D__HIPSYCL__ -std=c++17 -D__HIPSYCL_MULTIPASS_HIP_HEADER__="/tmp/hipSYCL-syclcc-9ypnu18w/__hipsycl_embedded_hip_kernels.hpp" -D__HIPSYCL_ENABLE_HIP_TARGET__ -x hip --cuda-host-only -isystem /home/aland/local/bin/../include/hipSYCL/std/hiplike -isystem /usr/lib/llvm-15/lib/clang/15.0.0/include/.. -U__FLOAT128__ -U__SIZEOF_FLOAT128__ -I/opt/rocm//include -I/opt/rocm//include --rocm-device-lib-path=/opt/rocm//amdgcn/bitcode --rocm-path=/opt/rocm/ -fhip-new-launch-api -mllvm -amdgpu-early-inline-all=true -mllvm -amdgpu-function-calls=false -D__HIP_ROCclr__ -D__HIPSYCL_MULTIPASS_CUDA_HEADER__="/tmp/hipSYCL-syclcc-9ypnu18w/__hipsycl_embedded_cuda_kernels.hpp" -D__HIPSYCL_ENABLE_CUDA_TARGET__ -D__HIPSYCL_ENABLE_OMPHOST_TARGET__ -I/usr/include -D_ENABLE_EXTENDED_ALIGNED_STORAGE -O3 test.cpp -o test -L/home/aland/local/bin/../lib/ -lhipSYCL-rt -Wl,-rpath=/home/aland/local/bin/../lib/ -Wl,-rpath=/opt/rocm//lib -Wl,-rpath=/opt/rocm//hip/lib -L/opt/rocm//lib -L/opt/rocm//hip/lib -lamdhip64 -Wl,-rpath=/usr/local/cuda/lib64 -L/usr/local/cuda/lib64 -lcudart -L/usr/lib/x86_64-linux-gnu -lboost_context -lboost_fiber -Wl,-rpath=/usr/lib/x86_64-linux-gnu

$ HIPSYCL_DRYRUN=1 syclcc -O3 --hipsycl-explicit-multipass --hipsycl-targets='hip:gfx1032' test.cpp -o test 
/usr/lib/llvm-15/bin/clang++ -I/home/aland/local/bin/../include/ -D__HIPSYCL__ -std=c++17 -isystem /home/aland/local/bin/../include/hipSYCL/std/hiplike -isystem /usr/lib/llvm-15/lib/clang/15.0.0/include/.. -U__FLOAT128__ -U__SIZEOF_FLOAT128__ -I/opt/rocm//include -I/opt/rocm//include --rocm-device-lib-path=/opt/rocm//amdgcn/bitcode --rocm-path=/opt/rocm/ -fhip-new-launch-api -mllvm -amdgpu-early-inline-all=true -mllvm -amdgpu-function-calls=false -D__HIP_ROCclr__ -x hip -D__HIPSYCL_ENABLE_HIP_TARGET__ -D__HIPSYCL_CLANG__ -fplugin=/home/aland/local/bin/../lib/libhipSYCL_clang.so --cuda-device-only --cuda-gpu-arch=gfx1032 -c -o /tmp/hipSYCL-syclcc-vv8jw2m7/hipfb-gfx1032.hipsycl_kernels -fpass-plugin=/home/aland/local/bin/../lib/libhipSYCL_clang.so -O3 test.cpp
/usr/lib/llvm-15/bin/clang++ -I/home/aland/local/bin/../include/ -D__HIPSYCL__ -std=c++17 -D__HIPSYCL_MULTIPASS_HIP_HEADER__="/tmp/hipSYCL-syclcc-vv8jw2m7/__hipsycl_embedded_hip_kernels.hpp" -D__HIPSYCL_ENABLE_HIP_TARGET__ -x hip --cuda-host-only -isystem /home/aland/local/bin/../include/hipSYCL/std/hiplike -isystem /usr/lib/llvm-15/lib/clang/15.0.0/include/.. -U__FLOAT128__ -U__SIZEOF_FLOAT128__ -I/opt/rocm//include -I/opt/rocm//include --rocm-device-lib-path=/opt/rocm//amdgcn/bitcode --rocm-path=/opt/rocm/ -fhip-new-launch-api -mllvm -amdgpu-early-inline-all=true -mllvm -amdgpu-function-calls=false -D__HIP_ROCclr__ -D__HIPSYCL_ENABLE_OMPHOST_TARGET__ -I/usr/include -D_ENABLE_EXTENDED_ALIGNED_STORAGE -O3 test.cpp -o test -L/home/aland/local/bin/../lib/ -lhipSYCL-rt -Wl,-rpath=/home/aland/local/bin/../lib/ -Wl,-rpath=/opt/rocm//lib -Wl,-rpath=/opt/rocm//hip/lib -L/opt/rocm//lib -L/opt/rocm//hip/lib -lamdhip64 -L/usr/lib/x86_64-linux-gnu -lboost_context -lboost_fiber -Wl,-rpath=/usr/lib/x86_64-linux-gnu
```

hipSYCL configuration is the default:

```console
$ syclcc --hipsycl-version
syclcc [hipSYCL compilation driver], Copyright (C) 2018-2022 Aksel Alpay and the hipSYCL project
  hipSYCL version: 0.9.3
  Installation root: /home/aland/local
  Plugin LLVM version: 15, can accelerate CPU: True
  Available runtime backends:
     librt-backend-cuda.so
     librt-backend-omp.so
     librt-backend-hip.so


Full configuration [can be overriden using environment variables or command line arguments]:
    version-major: 0
    version-minor: 9
    version-patch: 3
    plugin-llvm-version-major: 15
    plugin-with-cpu-acceleration: true
    default-clang: /usr/lib/llvm-15/bin/clang++
    default-nvcxx: (unconfigured)
    default-platform: cuda
    default-cuda-path: /usr/local/cuda
    default-gpu-arch: (unconfigured)
    default-cpu-cxx: /usr/bin/clang++-15
    default-rocm-path: /opt/rocm/
    default-use-bootstrap-mode: false
    default-is-dryrun: false
    default-use-accelerated-cpu: true
    default-clang-include-path: /usr/lib/llvm-15/lib/clang/15.0.0/include/..
    default-sequential-link-line: -L/usr/lib/x86_64-linux-gnu -lboost_context -lboost_fiber -Wl,-rpath=/usr/lib/x86_64-linux-gnu
    default-sequential-cxx-flags: -I/usr/include -D_ENABLE_EXTENDED_ALIGNED_STORAGE
    default-omp-link-line: -L/usr/lib/x86_64-linux-gnu -lboost_context -lboost_fiber -Wl,-rpath=/usr/lib/x86_64-linux-gnu -fopenmp
    default-omp-cxx-flags: -I/usr/include -fopenmp -D_ENABLE_EXTENDED_ALIGNED_STORAGE
    default-rocm-link-line: -Wl,-rpath=$HIPSYCL_ROCM_PATH/lib -Wl,-rpath=$HIPSYCL_ROCM_PATH/hip/lib -L/opt/rocm//lib -L/opt/rocm//hip/lib -lamdhip64
    default-rocm-cxx-flags: -isystem $HIPSYCL_PATH/include/hipSYCL/std/hiplike -isystem /usr/lib/llvm-15/lib/clang/15.0.0/include/.. -U__FLOAT128__ -U__SIZEOF_FLOAT128__ -I$HIPSYCL_ROCM_PATH/include -I$HIPSYCL_ROCM_PATH/include --rocm-device-lib-path=$HIPSYCL_ROCM_PATH/amdgcn/bitcode --rocm-path=$HIPSYCL_ROCM_PATH -fhip-new-launch-api -mllvm -amdgpu-early-inline-all=true -mllvm -amdgpu-function-calls=false -D__HIP_ROCclr__
    default-cuda-link-line: -Wl,-rpath=$HIPSYCL_CUDA_LIB_PATH -L$HIPSYCL_CUDA_LIB_PATH -lcudart
    default-cuda-cxx-flags: -U__FLOAT128__ -U__SIZEOF_FLOAT128__ -isystem $HIPSYCL_PATH/include/hipSYCL/std/hiplike
    default-is-explicit-multipass: false
    default-save-temps: false
```

Nothing interesting in runtime debug output:

```
[hipSYCL Info] kernel_cache: Querying kernel cache, having 2 code objects.
[hipSYCL Info] kernel_cache: Requested object not hooked up to kernel index, attempting full cache search.
[hipSYCL Info] kernel_cache: cache MISS; constructing new object for kernel index 0 and backend kernel name _Z16__hipsycl_kernelIN7hipsycl4glue20complete_kernel_nameIZZ11run_kernelsRKNS0_4sycl6deviceEENKUlRNS3_7handlerEE_clES8_E7KernelAEEEvv
[hipSYCL Error] from /home/aland/hipSYCL/src/runtime/hip/hip_code_object.cpp:125 @ build(): hip_executable_object: could not create module (error code = HIP:218)
[hipSYCL Error] from /home/aland/hipSYCL/include/hipSYCL/runtime/kernel_cache.hpp:291 @ get_or_construct_code_object_impl(): kernel_cache: code object creation has failed
[hipSYCL Error] from /home/aland/hipSYCL/src/runtime/hip/hip_queue.cpp:522 @ submit_kernel_from_code_object(): hip_queue: Code object construction failed
```

**Comment by illuhad** [comment]
Does it work if you remove the `--hipsycl-explicit-multipass` flag? It should automatically select a working configuration without it.

Does it work if you remove the `--hipsycl-explicit-multipass` flag and only set one of the two to explicit multipass, i.e 
* `--hipsycl-targets="hip.explicit-multipass:gfx1032;cuda.integrated-multipass:sm_86"`
* `--hipsycl-targets="hip.integrated-multipass:gfx1032;cuda.explicit-multipass:sm_86"`

Can you try dumping the HCF objects using the runtime environment variable `HIPSYCL_HCF_DUMP_DIRECTORY` and setting it to some directory where it can dump the HCF code. If both backends are in explicit multipass, you should get two HCF objects (one for CUDA, one for HIP) that we can inspect with `hipsycl-hcf-tool` to validate them. The error code indicates that the data that HIP gets is invalid.

I'm on vacation at the moment and cannot spend a lot of time debugging this unfortunately.

**Comment by illuhad** [comment]
Ok, I think I found the issue: https://github.com/illuhad/hipSYCL/blob/989026577b62409139c856491e9d67707448780a/include/hipSYCL/glue/generic/hiplike/hiplike_kernel_launcher.hpp#L946

This assumes that only one multipass is active at the same time. We cannot use macros here and should instead the `Backend_id` template argument. The current code will always pick the CUDA HCF object first if both CUDA and HIP are available.

I tested one in explicit, one in integrated and also both in explicit multipass, but it's possible that in the case where both are in explicit multipass I only tried running it on NVIDIA -.-
At least it should be an easy fix :)

**Comment by al42and** [comment]
> Does it work if you remove the `--hipsycl-explicit-multipass flag`? It should automatically select a working configuration without it.

Nope, without any flags explicit multipass gets enabled automatically and I get the same error.

> `--hipsycl-targets="hip.explicit-multipass:gfx1032;cuda.integrated-multipass:sm_86"`
> `--hipsycl-targets="hip.integrated-multipass:gfx1032;cuda.explicit-multipass:sm_86"`

Both work.

**Comment by al42and** [comment]
The fix in #831 solves the problem. Thank you, @illuhad, and sorry for interrupting your vacation
